packages/matter.js/src/cluster/Cluster.ts

```diff
@@ -22,58 +22,487 @@ export const enum AccessLevel {
 
-export const Command = <RequestT, ResponseT>(requestId: number, requestSchema: TlvSchema<RequestT>, responseId: number, responseSchema: TlvSchema<ResponseT>): Command<RequestT, ResponseT> => ({ optional: false, requestId, requestSchema, responseId, responseSchema });
-export const OptionalCommand = <RequestT, ResponseT>(requestId: number, requestSchema: TlvSchema<RequestT>, responseId: number, responseSchema: TlvSchema<ResponseT>): OptionalCommand<RequestT, ResponseT> => ({ optional: true, requestId, requestSchema, responseId, responseSchema });
+interface ConditionalCommandOptions<F extends BitSchema> {
+    optionalIf?: ConditionalFeatureList<F>,
+    mandatoryIf?: ConditionalFeatureList<F>
+}
+
+export const Command = <RequestT, ResponseT>(
+    requestId: number,
+    requestSchema: TlvSchema<RequestT>,
+    responseId: number,
+    responseSchema: TlvSchema<ResponseT>
+): Command<RequestT, ResponseT> => ({
+    optional: false,
+    requestId,
+    requestSchema,
+    responseId,
+    responseSchema
+});
+
+export const OptionalCommand = <RequestT, ResponseT>(
+    requestId: number,
+    requestSchema: TlvSchema<RequestT>,
+    responseId: number,
+    responseSchema: TlvSchema<ResponseT>,
+): OptionalCommand<RequestT, ResponseT> => ({
+    optional: true,
+    requestId,
+    requestSchema,
+    responseId,
+    responseSchema
+});
+
+export const ConditionalCommand = <RequestT, ResponseT, F extends BitSchema>(
+    requestId: number,
+    requestSchema: TlvSchema<RequestT>,
+    responseId: number,
+    responseSchema: TlvSchema<ResponseT>,
+    { optionalIf, mandatoryIf }: ConditionalCommandOptions<F>
+): ConditionalCommand<RequestT, ResponseT, F> => ({
+    optional: true,
+    optionalIf,
+    mandatoryIf,
+    requestId,
+    requestSchema,
+    responseId,
+    responseSchema
+});
 
 /* Interfaces and helper methods to define a cluster event */
 export const enum EventPriority {
     Critical,
     Info,
     Debug,
 }
-export interface Event<T> { id: number, schema: TlvSchema<T>, priority: EventPriority, optional: boolean }
-export interface OptionalEvent<T> extends Event<T> { optional: true }
-export const Event = <FT extends TlvFields>(id: number, priority: EventPriority, data: FT = <FT>{}): Event<TypeFromFields<FT>> => ({ id, schema: TlvObject(data), priority, optional: false });
-export const OptionalEvent = <FT extends TlvFields>(id: number, priority: EventPriority, data: FT = <FT>{}): OptionalEvent<TypeFromFields<FT>> => ({ id, schema: TlvObject(data), priority, optional: true });
+
+export interface Event<T> {
+    id: number,
+    schema: TlvSchema<T>,
+    priority: EventPriority,
+    optional: boolean
+}
+
+interface ConditionalEventOptions<F extends BitSchema> {
+    optionalIf?: ConditionalFeatureList<F>,
+    mandatoryIf?: ConditionalFeatureList<F>
+}
+
+export interface OptionalEvent<T> extends Event<T> {
+    optional: true
+}
+
+export interface ConditionalEvent<T, F extends BitSchema> extends OptionalEvent<T> {
```

Comment:
I think this can just extend OptionalEvents<T> & ConditionalEventOptions<T> (or ElementOptions<T>) and have an empty body?